src/WorkItemMigrator/JiraExport/JiraItem.cs

```diff
@@ -66,10 +67,29 @@ private static List<JiraRevision> BuildRevisions(JiraItem jiraItem, IJiraProvide
                 Dictionary<string, object> fieldChanges = new Dictionary<string, object>(StringComparer.InvariantCultureIgnoreCase);
 
                 var items = change.SelectTokens("$.items[*]")?.Cast<JObject>()?.Select(i => new JiraChangeItem(i));
-
                 foreach (var item in items)
                 {
```

Comment:
I dont follow the logic behined this. Can you explain it ?

Reply:
Say we start this method parsing ISSUE-2, that looks like this:
```{
  "expand": "renderedFields,names,schema,operations,editmeta,changelog,versionedRepresentations",
  "id": "59698",
  "self": "https://jiraserver/rest/api/2/issue/59698",
  "key": "ISSUE-2",
  "fields": {
    "issuetype": {
    },
    "parent": {
      "id": "57462",
      "key": "ISSUE-1"
    }
  },
  "timespent": null,
  "project": {
  }
}
```
ISSUE-2 has ISSUE-1 as parent. Let's also say this parent was added not in the initial revision, but in a later revision. So that the changelog looks like this:
```
{
  "expand": "renderedFields,names,schema,operations,editmeta,changelog,versionedRepresentations",
  "id": "59698",
  "self": "https://jiraserver/rest/api/2/issue/59698",
  "key": "ISSUE-2",
  "fields": { "created": "2020-01-16T10:55:48.737+0000" },
  "renderedFields": { "created": "16/Jan/20 10:55 AM" },
  "changelog": {
    "startAt": 0,
    "maxResults": 1,
    "total": 1,
    "histories": [
      {
        "id": "729344",
        "items": [
          {
            "field": "Parent",
            "fieldtype": "jira",
            "from": null,
            "fromString": null,
            "to": "57462",
            "toString": "ISSUE-1"
          }
        ]
      }
    ]
  }
}
```

So the logic here, is to loop over all changes in the changelog ("histories[]") and undo every changed field ("Items[]").  That means we will record the parent link as a revision with `fieldChanges["parent"] = item.ToString;` and we will also remove it from the issue fields with: `fields["parent"] = item.FromString;`

With the field now removed from the issue, we end up with the correct initial revision, without the parent link. And a correct revision 1 where a parent link is added.

Note that the changelog refers the the field as "Parent" so the standard undo logic doesn't work because the field name is "parent". I've fixed this by duplicating the code for this case, and hardcoding this special case. Admittedly, not the cleanest solution, but it works. What we need here is a mapping between how the Changelog refers to a field and the actual fieldname in the issue.

The same applies to "Epic Link"  except here we used the Epic Link configurable custom field. Because the changelog refers to this field as "Epic Link" but the issue field might be "customfield_10700" .

The unittests also test for this case.

Reply:
The rest API give the changelog and the issue with the changelong in different order. Now that in another PR you already are sorting the order by ID that should be fixed already. 
Also I dont see how the link can be in the initial revision because running a query to get change log with the same setup you said i dont see that being the case, henc my questions :
Here
```
HTTP/1.1 200 OK
Date: Wed, 26 Oct 2022 12:54:56 GMT


{
  "self": "https://ourOrg.atlassian.net/rest/api/2/issue/SCRUM-76/changelog?maxResults=100&startAt=0",
  "maxResults": 100,
  "startAt": 0,
  "total": 3,
  "isLast": true,
  "values": [
    {
      "id": "10866",
      "author": {
        "self": "https://ourOrg.atlassian.net/rest/api/2/user?accountId=5f8d85fe9c31840076d047fa",
        "accountId": "5f8d85fe9c31840076d047fa",
        "emailAddress": "madis.koosaar@solidify.dev",
        "avatarUrls": {
          "48x48": "https://secure.gravatar.com/avatar/fa25ab2ab892cf05eec8eb5c68d19323?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMK-2.png",
          "24x24": "https://secure.gravatar.com/avatar/fa25ab2ab892cf05eec8eb5c68d19323?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMK-2.png",
          "16x16": "https://secure.gravatar.com/avatar/fa25ab2ab892cf05eec8eb5c68d19323?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMK-2.png",
          "32x32": "https://secure.gravatar.com/avatar/fa25ab2ab892cf05eec8eb5c68d19323?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMK-2.png"
        },
        "displayName": "Madis Koosaar",
        "active": true,
        "timeZone": "Europe/Berlin",
        "accountType": "atlassian"
      },
      "created": "2022-10-26T14:46:14.271+0200",
      "items": [
        {
          "field": "assignee",
          "fieldtype": "jira",
          "fieldId": "assignee",
          "from": null,
          "fromString": null,
          "to": "5f8d85fe9c31840076d047fa",
          "toString": "Madis Koosaar",
          "tmpFromAccountId": null,
          "tmpToAccountId": "5f8d85fe9c31840076d047fa"
        }
      ]
    },
    {
      "id": "10867",
      "author": {
        "self": "https://ourOrg.atlassian.net/rest/api/2/user?accountId=5f8d85fe9c31840076d047fa",
        "accountId": "5f8d85fe9c31840076d047fa",
        "emailAddress": "madis.koosaar@solidify.dev",
        "avatarUrls": {
          "48x48": "https://secure.gravatar.com/avatar/fa25ab2ab892cf05eec8eb5c68d19323?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMK-2.png",
          "24x24": "https://secure.gravatar.com/avatar/fa25ab2ab892cf05eec8eb5c68d19323?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMK-2.png",
          "16x16": "https://secure.gravatar.com/avatar/fa25ab2ab892cf05eec8eb5c68d19323?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMK-2.png",
          "32x32": "https://secure.gravatar.com/avatar/fa25ab2ab892cf05eec8eb5c68d19323?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMK-2.png"
        },
        "displayName": "Madis Koosaar",
        "active": true,
        "timeZone": "Europe/Berlin",
        "accountType": "atlassian"
      },
      "created": "2022-10-26T14:46:52.972+0200",
      "items": [
        {
          "field": "IssueParentAssociation",
          "fieldtype": "jira",
          "from": null,
          "fromString": null,
          "to": "10069",
          "toString": "SCRUM-42"
        }
      ]
    },
    {
      "id": "10869",
      "author": {
        "self": "https://ourOrg.atlassian.net/rest/api/2/user?accountId=5f8d85fe9c31840076d047fa",
        "accountId": "5f8d85fe9c31840076d047fa",
        "emailAddress": "madis.koosaar@solidify.dev",
        "avatarUrls": {
          "48x48": "https://secure.gravatar.com/avatar/fa25ab2ab892cf05eec8eb5c68d19323?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMK-2.png",
          "24x24": "https://secure.gravatar.com/avatar/fa25ab2ab892cf05eec8eb5c68d19323?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMK-2.png",
          "16x16": "https://secure.gravatar.com/avatar/fa25ab2ab892cf05eec8eb5c68d19323?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMK-2.png",
          "32x32": "https://secure.gravatar.com/avatar/fa25ab2ab892cf05eec8eb5c68d19323?d=https%3A%2F%2Favatar-management--avatars.us-west-2.prod.public.atl-paas.net%2Finitials%2FMK-2.png"
        },
        "displayName": "Madis Koosaar",
        "active": true,
        "timeZone": "Europe/Berlin",
        "accountType": "atlassian"
      },
      "created": "2022-10-26T14:46:53.053+0200",
      "items": [
        {
          "field": "Epic Link",
          "fieldtype": "custom",
          "fieldId": "customfield_10008",
          "from": null,
          "fromString": null,
          "to": "10069",
          "toString": "SCRUM-42"
        }
      ]
    }
  ]
}

```

Reply:
You used a different API here https://ourOrg.atlassian.net/rest/api/2/issue/SCRUM-76/changelog because the code uses https://ourOrg.atlassian.net/rest/api/2/issue/SCRUM-76?expand=changelog,renderedFields&fields=created it seems slightly syntactically different. I've never seen field "IssueParentAssociation" before, is that the same in the other API?

Anyway, in your example above, the Epic Link is added as the 3rd change to the issue. With the current master branch I would expect this Epic Link to show up in revision index 0 in the exported json file. With this PR applied I would expect it to show up in revision index 3 (where it should be).

Reply:
Indeed I sent the wrong output as an example. There are two endpoints, either getting the changelog directly or getting the issue with expanded parameters. 

Still, the issue is that with getting the change long from `GET /rest/api/3/issue/{issueIdOrKey}` endpoint the order of the changelog is ordered from newst to oldest as Jira shows too. That is the reason why you are saying that the initial revision has the epic link that it should not. Just fixing the order should already solve the problem.

Reply:
The changelog order issue is related and required for a correct initial state, true, but it's not the same issue. The issue here is that `fields.Remove();` is never correctly called for an Epic Field or a Parent, and thus the initial revision is incorrect. This happens independently of the order in which changes are processed.

For a parent link currently fields.Remove("Parent") is called, where it should be fields.Remove("parent").
For an epic link currently fields.Remove("Epic Link") is called, where is should be fields.Remove("customfield_xxxxx").

Reply:
I did a test run with your code changes and without. For both runs, I got the exact same results that the Epic-Link was in the initial revision. Check my files 
[Archive.zip](https://github.com/solidify/jira-azuredevops-migrator/files/9878318/Archive.zip)

Or Am I missing something? 

One more thing I don't follow is that even if the link is in the initial revision when you import the issues will be imported in numerical order from 1-999, so in the case where you have issue-1 as epic and issue-2 as story the link created in AzureDevops will be fine, even the other way around because in Azure DevOps the links should be bi-directional and there should be a reference on both exported issues.

Thanks for bearing with me with all the questions :)

Reply:
Awesome, thanks for trying it out. The results are as expected. In file SCRUM-76-with-changes, you have the Link as revision index 3. That's where it should be and means this PR works as intended.

![image](https://user-images.githubusercontent.com/1309462/198298704-21238ad0-3ecf-48df-8a96-7dae3326f144.png)


 Indeed, it's also still in the initial revision as well, that's probably because of the changelog ordering. To be sure, you can apply that change too, and the result should be 100% correct.

About your second point, you are probably right that this results in a correct data migration either way. I'm a little fuzzy about this, let's play out a simple example.

```
01:00h; revision 0; Create SCRUM-1
02:00h; revision 0; Create SCRUM-2
03:00h: revision 1; Create link from SCRUM-2 to SCRUM-1
```

The import ordering is by date. Currently this means:
First SCRUM-1 rev 0 -> throws an error on the link, because SCRUM-2 does not exist
Then SCRUM 2 rev 0 -> link to SCRUM-1 succeeds
Then SCRUM 2 rev 1 -> empty revision
Then SCRUM 1 rev 1 -> empty revision

With the PR this means:
First SCRUM-1 rev 0 -> Creates workitem
Then SCRUM 2 rev 0 -> Creates workitem
Then SCRUM 2 rev 1 -> Create link to SCRUM-1
Then SCRUM 1 rev 1 -> Throws an error because the link already exists

That means we still have an error. But at least this is more correct from sequence of events perspective. We can fix the remaining error in a different PR.

No worries, thank you (and @Alexander-Hjelm ) for the active maintainership!

Reply:
Just to note I did have the ordering fix in it when creating the SCRUM-76-withchange output.

Reply:
Did you also configure "epic-link-field" in the import.json? Might be that's why it wasn't able to remove the correct field.

Reply:
Yes that was configured as well.

Reply:
@moerketh, what is the status here? Have you and madis been able to agree on this issue? Or are there any todos left?

Reply:
What happened here is, when adding an Epic Link in Jira Cloud, it's silently also adding a parent link at the same time with fieldname "IssueParentAssociation" in the changelog and fieldname "parent" on the issue. Plus, it's actually setting the Parent field twice. See:

![image](https://user-images.githubusercontent.com/1309462/200019579-7337a315-bb6d-45f9-a9ff-a984dc53b5d1.png)

Because the fieldnames between changelog and issue do not match this Parent link stays in the initial revision. And only the Epic Link has been moved to revision 3.

Jira Server 8 does not have the same behaviour, an Epic Link just sets the Epic Link field. That's why this change did not produce the same result for Cloud as it does for Server.

I'm thinking the easiest fix is to add support for the "IssueParentAssociation" changelog entry for Jira Cloud.

Reply:
@madkoo I've added support for the "IssueParentAssociation" in Jira Cloud. Can you give this another try?

Reply:
Looks fine now